### PR TITLE
[Concurrent Low-Code] ConcurrentDeclarativeSource class that low-code connectors can inherit from to uptake Concurrent CDK

### PR DESCRIPTION
**Pull request by brianjlai**
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/9710

## What

Adds the new `ConcurrentDeclarativeSource` class which serves as the way we adapt the existing YamlDeclarativeSource used by all low-code connectors into being runnable within the Concurrent CDK framework. This PR combines all the other previous units of work so that low-code streams are translated into concurrent `DefaultStream` instances.

Another big aspect of this review is how we gate the streams that will run concurrently:

- Only incremental non-substreams will run concurrently and we do so by inspecting for the DatetimeBasedCursor
- We cannot migrate full refresh streams to concurrent because concurrent CDK does not support RFR so we would be regressing in behavior.
- We cannot migrate substreams because concurrent CDK does not support substreams

## How

The overall design is predicated on introducing a new class `ConcurrentDeclarativeSource` which behaves as a kind of adapter between the existing entrypoint.py that all syncs are triggered from and the `ConcurrentSource` which is responsible for running certain streams using the Concurrent CDK engine.

The `ConcurrentDeclarativeSource` inherits the ManifestDeclarativeSource so that we can reuse the logic to parse a manifest into low-code runtime components and allow the inspection of it's components to decide whether it can be run concurrently or synchronously.

The last big part of the code is this puts into place the logic to transform a low-code stream's `DatetimeBasedCursor` into a `ConcurrentCursor`. The reason why we need to do this is that the interfaces for a low-code cursor and concurrent cursor differ in a few specific ways and trying to make them both fit the same interface created a frankenstein class that proved to be even more unwieldly. In prior PRs, see [45413](https://github.com/airbytehq/airbyte/pull/45413), it was determined that there was feature parity so now we perform the transformation and supply it to the concurrent engine to handle date window partitioning and state management.

Something else important to note is that there are some specific cases where an incremental stream cannot be run as a concurrent source. Since we introduced the language, we've allow `stream_state` to be a valid interpolation for various components. However, because partitions can be run in any order and complete at anytime, `stream_state` managed by the ConcurrentCursor is not a thread-safe value anymore (vs when it was managed sequentially). I inspected the schema and our repo for it's usage. For streams using `stream_state` in an unsafe way, we make it a synchronous stream, but we should fix those connectors to use the thread safe `stream_interval` and ultimately get rid of the extra code later.

### Short term how we enable this

I've included two examples of how connectors can uptake concurrent processing. They are the same and will be deleted before merging.

The two things that need to be changed are:

- `run.py` - Our previous design for connectors did not take any arguments passed to the connector from the platform. This is a significant limitation because the concurrent framework is entirely based around instantiating things like cursors up front before performing a read. I haven't found a great way to avoid changing this as this is a limitation of the Concurrent CDK
- `source.py` - Once `run.py` is updated to pass in the various operation arguments like state, config, and catalog, we need to pass them to the `ConcurrentDeclarativeSource` constructor

## Review guide

1. `concurrent_declarative_source.py`
2. `datetime_based_cursor.py`
3. `adapter.py`
4. `datetime_stream_state_converter.py`
5. `yaml_declarative_source.py`
6. See either `source-sentry` or `source-amplitude`

## User Impact

This is considered a breaking CDK change because connectors will need to follow the included migration guide to update a connectors `run.py` and `source.py` files

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

Yes, because this isn't release yet. However, this does pose a risk once we move a connector to concurrent because once we start emitting the new state format, then it is much harder to go backward since the connector cannot process concurrent state. See my comment in the code for more

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:157]: By disabling `is_sequential_state`, we will automatically move connectors from their sequential state to per-partition concurrent state. This is the ideal end goal we want, however, if we were to revert the connector to a previous version that doesn't use the concurrent CDK, the state message would not be compatible.

It feels like to de-risk this a little bit on the early release, we should set this to true so we accept sequential state (i.e. `{"updated_at": "2024-12-12"}` and also emit it back to the platform as such

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_source.py
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_source.py:38]: For non concurrent low-code sources, these are equivalent, but we override this implementation in `ConcurrentDeclarativeSource` to create a single list of both concurrent and synchronous sources so that we properly generate catalogs and other things

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py
Comment (natikgadzhi) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:31]: Teach me your ways, what is the difference betweeen streams and all_streams? (ignore if it's in this PR, reading through it)

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/availability_strategy.py
Comment (natikgadzhi) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/availability_strategy.py:70]: @brianjlai probably better to call out that it should not be used at all, if we're ripping out availability strategies over mid-long term?

**Review by natikgadzhi — COMMENTED** [review]
Got distracted, just posting this so comments don't get lost, will read thoroughly later

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:31]: good question, I mention it in a comment: https://github.com/airbytehq/airbyte/pull/46662/files#r1792816384

But the reason why we can't just rewrite the `streams()` method is because within the existing Python CDK `core.py`, when processing synchronous streams, we invoke the `streams()` method and in that context we don't want to return the concurrent streams that aren't compatible in that are of code.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/availability_strategy.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/availability_strategy.py:70]: ah yeah i think i carried that over from a merge of serhii's work which deprecated availability strategy in concurrent. I'll update this to say do not use

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (lazebnyi) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py]: do you think we could add a mechanism to handle this situation by using the lowest value to set the sequential state if we revert the changes?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py
Comment (lazebnyi) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:31]: As we discussed earlier, it’s preferable to use the stream method and condition the behavior accordingly. This approach adds some complexity, but it provides a tradeoff by allowing simpler modifications later. With this setup, when the core will be able to handle concurrent streams, we’ll get a stream generation interface for free.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (pnilan) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:394]: nit: It seems like it would be more memory efficient to directly iterate over the generated slices, is there a specific reason for saving to a list?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (lazebnyi) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:85]: I think it would be better to move the value from the code into a class variable may be to improve readability

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (lazebnyi) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:138]: Do you think the name `_group_streams` would be more informative?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (lazebnyi) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:390]: For current implementation of datetime cursor `self._slice_boundary_fields` never has none value

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/run.py
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/run.py:17]: I think we need to mention in PR description clearly that this is will be a breaking change. And add info about this to cdk-migration file.

**Review by lazebnyi — CHANGES_REQUESTED** [review]
Overall, the implementation looks great—nice work! However, we still need to make a few updates. After that I can approve.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (pnilan) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:211]: I don't follow this -- Is there somewhere we would be serializing a datetime object?

**Review by pnilan — COMMENTED** [review]
A couple minor comments. I'll defer to Serhii for approval.

Also can you run regression tests w/ one/two of the test connectors?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:138]: i'm fine with `_group_streams`

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/run.py
Comment (brianjlai) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/run.py:17]: yep that is the plan, I wrote up a migration guide that will be included in the next commit I push explaining what needs to change  in `run.py` and `source.py`.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:144]: it is. i'm removing this thank you

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:223]: good eyes!

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:394]: correct. I had originally added this for debugging to see the entire set of slices easier, but you are correct this should be iterable

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:211]: i'll double check, but I think we needed this serializer deeper in our code, potentially in how we emit state back out. I'll reconfirm this as I work through @lazebnyi comment above.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py]: yep this is addressed in my latest commit using the optional param `include_concurrent_streams`

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:269]: Do we have documentation somewhere on things we need to do to fully migrate to the one CDK? If not, I can start one and this will be part of the list.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:316]: Why not `stream.get_cursor()` here? I fear that this will not work for `AsyncRetriever`

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:309]: The concern I have with this method is that we are coupling the creation of the concurrent cursor with code that should eventually be removed i.e. we do (declarative -> low-code component -> concurrent) instead of (declarative -> concurrent). Once we will remove the low-code components, we will have to re-implement this logic. Should we instead rely on `airbyte_cdk.sources.declarative.models.declarative_component_schema.DatetimeBasedCursor`?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:211]: I'm also curious about this. From my understanding, this would mean that CursorPartitionGenerator would create slices with datetime within them but this is not what I see.

File: airbyte-cdk/python/cdk-migrations.md
Comment (maxi297) [inline airbyte-cdk/python/cdk-migrations.md:3]: Can we validate with @bnchrch what is the impact on the manifest-only sources and what is the migration path for them?

**Review by maxi297 — COMMENTED** [review]
Here is a first pass of review. I'm very happy with the so many edge cases you've caught during this. 

I have concerns but all of them are opened to discussion and I'm willing to tackle some of them myself if it can help.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py:211]: @lazebnyi @pnilan @maxi297 to close the loop on this one. I think what originally happened was i wrote this in when i was first testing because we were getting the datetime object from the ConcurrentCursor and it would fail trying to serialize it.

However, later after cleaning up the code and fixing edge cases, I addressed the serialization by converting the datetime into the correct output string format with the correct precision in the generate() function [here](https://github.com/airbytehq/airbyte/pull/46662/files#diff-93127bface0b323fe43b21cdb8fb14493dd465995b085a4f81647f3697930bddR396-R399) . And since this was now already a string we don't need to serialize it. 

I'll get rid of this code as it's not actually used anymore and we're applying the correct precision based on the cursor definition.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:286]: This is a good idea. i'll add a log to each of these scenarios

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:269]: We do not have it documented at a high level as far as I am aware. However, in my writeup for the breaking change in `cdk-migrations.md`, I've listed all the connectors in the repo that use `stream_state` in a non-thread safe way and what we can do to fix them.

But at a higher level, we should have a line item for the unification of our CDKs. Basically we should update our schemas to not use `stream_state` as an interpolation context and remove it wherever it's used from the Retriever and below (although I don't think it's actually used above that)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:316]: yeah this is just an oversight on my part, i've just got rid of this whole method and just call self.get_cursor() from above instead

File: airbyte-cdk/python/cdk-migrations.md
Comment (brianjlai) [reply airbyte-cdk/python/cdk-migrations.md:3]: yep, i'll add a note here: But it sounds like `source-declarative-manifest` is pinned to an explicit, version and as for connector builder, that one actually doesn't use `YamlDeclarativeSource` it uses `SourceDeclarativeManifest` which doesn't contain breaking changes. I think it's up for debate whether that should switch over, but for now it would be non breaking for there

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:309]: That is a fair question. I think this can get a little bit awkward because we need to sort of extract the DatetimeBasedCursor model from the `_source_config` field and we lose the automatic parsing into interpolated fields and other things that we basically need to reimplement. But it does feel like the model is the more solid interface that we won't change. I will play around with this tomorrow and see if this can be done sanely

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py:31]: Proposed temporary fix: https://github.com/airbytehq/airbyte/pull/46995/files#r1806474429

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:144]: Proposed fix: https://github.com/airbytehq/airbyte/pull/46995

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py:268]: I'll take some time to improve previous work on documenting this [here](https://whimsical.com/untangling-rfr-concurrent-low-code-VbjEeGwy72UeDEQDSWiMQ3). Thanks for sharing this!

**Review by lazebnyi — APPROVED** [review]
LGTM 
Great job!

**Comment by brianjlai** [comment]
regression testing report:

source-chargebee:
- Running single threaded w/ no concurrency_level, but still running using the concurrent framework:
  - https://github.com/airbytehq/airbyte/actions/runs/11564546803/job/32189934680
  - one extra record, a little bit odd, but it's possible that we might get additional records since we no longer run sequentially
- Running with a concurrency level of 3 using the concurrent framework
  - https://github.com/airbytehq/airbyte/actions/runs/11565622640/job/32192921374
  - All records match. Many of `source-chargebee`'s streams are concurrent with time windows.
- Running a single stream `customer` without state to test date windows using concurrency of 3
  - https://github.com/airbytehq/airbyte/actions/runs/11526008466/job/32089390924
  - There is a exact match on record count. There are data mismatches, but I believe this could be a product of the entire sync no longer syncing sequentially so changing data across 50,000+ records is possible
- Running a single stream `customer` without state to test date windows with single level of concurrency
  - https://github.com/airbytehq/airbyte/actions/runs/11567312962/job/32197412341
  - Ran this test to verify that the above case where data changed across the 50,000+ records
  - Successful with no mismatches when run single-threaded on the concurrent framework

source-sentry:
- Running using full refresh:
  - https://github.com/airbytehq/airbyte/actions/runs/11566207300/job/32194470052
  - All records for the streams match being run as a data feed
  - The mismatched catalogs are a product of going to v5 of the CDK where `is_resumable` is added
  - The mismatched null values are because in v4 of the CDK we optimized output to no longer emit `null` fields. This is unrelated to my current change
- todo, re-run a sync using state once the target candidate has more records populated

overall analysis:
the two connectors test different features of low-code incremental syncing `source-sentry` is a single threaded data feed that is parallelized per stream. `source-chargebee` is a traditional windowed incremental sync where different partitions are run concurrently. So far I don't see other issues that would block release of the change and per @maxi297 final discussion and approval

**Review by maxi297 — APPROVED** [review]
I'm fine with those changes. Thanks for the thorough testing @brianjlai !

**Pull request merged**
